### ISSUE TITLE
Getting the car to wake using BLE

### ISSUE THREAD

**Issue by iainbullock** [iainbullock/tesla_ble_mqtt_docker#2]
Thanks @iainbullock! Looking great.
You mention in your docs that the `wake` command does not work. I managed to get it to work with the `-domain vcsec` parameter. I will try to contribute directly then.
Are you familiar with HA integrations? It seems that with this, we are not too far from deriving your project to make the integration work locally on the HA host 😀. It should be feasible, but so far I do not know how to access the BLE device in HAOS...

Anyways, this thread is getting very long and it seems that most of us have a workaround with FleetAPI (with increased polling intervals) + BLE for frequent commands when close to the car.

I suggest to **close this issue** as solved, marking Ian's repo as solution... @alandtse? At least until a more packaged solution raises with the full bundle HA extension: Fleet API + local BLE (on HA host) + remote BLE (on ESP or RPi), including some proximity/location sensors...

Maybe switch this discussion to Discord or anything less verbose on the Internet? I just initialized [this channel](https://discord.com/channels/1249735849080520825/1249735849080520828)

_Originally posted by @raphmur in https://github.com/alandtse/tesla/issues/961#issuecomment-2158589461_

**Comment by iainbullock** [comment]
@raphmur Any chance to can elaborate on the -domain vcsec parameter?

Thanks

**Comment by iainbullock** [comment]
I ran tesla-control --help, which gave me a clue:

```
Usage: tesla-control [OPTION...] COMMAND [ARG...]

Run tesla-control help COMMAND for more information. Valid COMMANDs are listed below.

 * Commands sent to a vehicle over the internet require a VIN and a token.
 * Commands sent to a vehicle over BLE require a VIN.
 * Account-management commands require a token.

Available OPTIONs:
  -ble
        Force BLE connection even if OAuth environment variables are defined
  -command-timeout duration
        Set timeout for commands sent to the vehicle. (default 5s)
  -connect-timeout duration
        Set timeout for establishing initial connection. (default 20s)
  -debug
        Enable verbose debugging messages
  -domain value
        Domains to connect to (can be repeated; omit for all)
  -key-file file
        A file containing private key. Defaults to $TESLA_KEY_FILE.
  -key-name name
        System keyring name for private key. Defaults to $TESLA_KEY_NAME.
  -keyring-debug
        Enable keyring debug logging
  -keyring-file-dir directory
        keyring directory for file-backed keyring types (default "~/.tesla_keys")
  -keyring-type type
        Keyring type (file|keyctl|pass). Defaults to $TESLA_KEYRING_TYPE.
  -session-cache file
        Load session info cache from file. Defaults to $TESLA_CACHE_FILE.
  -token-file File
        File containing OAuth token. Defaults to $TESLA_TOKEN_FILE.
  -token-name name
        System keyring name for OAuth token. Defaults to $TESLA_TOKEN_NAME.
  -vin string
        Vehicle Identification Number. Defaults to $TESLA_VIN.

Available COMMANDs:
  add-key                  Add PUBLIC_KEY to vehicle whitelist with ROLE and FORM_FACTOR
  add-key-request          Request NFC-card approval for a enrolling PUBLIC_KEY with ROLE and FORM_FACTOR
  auto-seat-and-climate    Turn on automatic seat heating and HVAC
  autosecure-modelx        Close falcon-wing doors and lock vehicle. Model X only.
  charge-port-close        Close charge port
  charge-port-open         Open charge port
  charging-schedule        Schedule charging to MINS minutes after midnight and enable daily scheduling
  charging-schedule-cancel Cancel scheduled charge start
  charging-set-amps        Set charge current to AMPS
  charging-set-limit       Set charge limit to PERCENT
  charging-start           Start charging
  charging-stop            Stop charging
  climate-off              Turn off climate control
  climate-on               Turn on climate control
  climate-set-temp         Set temperature (Celsius)
  drive                    Remote start vehicle
  flash-lights             Flash lights
  frunk-open               Open vehicle frunk. Note that there's no frunk-close command!
  get                      GET an owner API http ENDPOINT. Hostname will be taken from -config.
  honk                     Honk horn
  list-keys                List public keys enrolled on vehicle
  lock                     Lock vehicle
  media-set-volume         Set volume
  media-toggle-playback    Toggle between play/pause
  ping                     Ping vehicle
  post                     POST to ENDPOINT the contents of FILE. Hostname will be taken from -config.
  product-info             Print JSON product info
  remove-key               Remove PUBLIC_KEY from vehicle whitelist
  rename-key               Change the human-readable metadata of PUBLIC_KEY to NAME, MODEL, KIND
  seat-heater              Set seat heater at POSITION to LEVEL
  sentry-mode              Set sentry mode to STATE ('on' or 'off')
  session-info             Retrieve session info for PUBLIC_KEY from DOMAIN
  software-update-cancel   Cancel a pending software update
  software-update-start    Start software update after DELAY
  steering-wheel-heater    Set steering wheel mode to STATE ('on' or 'off')
  tonneau-close            Close Cybertruck tonneau.
  tonneau-open             Open Cybertruck tonneau.
  tonneau-stop             Stop moving Cybertruck tonneau.
  trunk-close              Closes vehicle trunk. Only available on certain vehicle types.
  trunk-move               Toggle trunk open/closed. Closing is only available on certain vehicle types.
  trunk-open               Open vehicle trunk. Note that trunk-close only works on certain vehicle types.
  unlock                   Unlock vehicle
  wake                     Wake up vehicle
  windows-close            Close all windows
  windows-vent             Vent all windows
```

I see there is now a wake command (it wasn't list in Shankar's blog). I will try that

I also see the -domain option, so I assume that's what you were referring to? I'm not sure what it actually does and whether it is needed?

**Comment by PedroKTFC** [comment]
I'm finding this discussion very interesting for various reasons. Many people seem to have a problem with the rate limit applied by Tesla although that isn't affecting me at the moment at least. However, the use of BLE to detect the car's presence is of interest as I have an electric gate that I want to open automatically when I get in the car to leave home and when I arrive home. Hitherto, as I mentioned before, I used a simple ESP32 "dongle" I plugged into my car's usb port. Pre-Tesla, the port would only be live when the car ignition is turned on. A simple ping every 1s from HA would detect the dongle when it's live and in range and I'd use an automation to open the gate whenever the dongle is detected. Unfortunately, this approach doesn't work with my Tesla as it seems the usb ports can be live at almost random times (not just when someone's in the car).

So I wonder if I can use iBeacons? I confess to knowing very little about them and have a few questions I'd be really grateful if anyone can answer:
- Is there some good documentation about iBeacons (and in HA)? Also, how to use them with a Tesla.
- How does a Tesla use iBeacons? Does a Tesla continuously broadcast its presence using iBeacons and what information is in such broadcasts? If not continuously, what triggers its iBeacons? 
- I run HA on a RPi 4 and have a RaspBee II for ZigBee so I don't think using the Pi for BLE would be sensible, what would be the best alternative?

I know I'm asking for a repeat of some of the discussions above and in other threads but it's really hard to follow the approaches. It would be great to have a single "how to" article that could be updated as people do things differently. I'll certainly write up what I do and post it on my substack (free!) but at the moment I don't know where to start!

**Comment by raphmur** [comment]
@iainbullock yes I found this special argument to be used on another forum, although I did not investigate what it does as "it works".

In your package shell simply add a special case for "wake" and change the simple $msg to "-domain vcsec $msg".

Indeed the --help gives lots of other possible commands. Basically LTE commands could really replace all the commands that fail through the Fleet API. Even the remote start, which Can be considered a security hazard in case of network/RPi hack.

**Comment by iainbullock** [comment]
> I'm finding this discussion very interesting for various reasons. Many people seem to have a problem with the rate limit applied by Tesla although that isn't affecting me at the moment at least. However, the use of BLE to detect the car's presence is of interest as I have an electric gate that I want to open automatically when I get in the car to leave home and when I arrive home. Hitherto, as I mentioned before, I used a simple ESP32 "dongle" I plugged into my car's usb port. Pre-Tesla, the port would only be live when the car ignition is turned on. A simple ping every 1s from HA would detect the dongle when it's live and in range and I'd use an automation to open the gate whenever the dongle is detected. Unfortunately, this approach doesn't work with my Tesla as it seems the usb ports can be live at almost random times (not just when someone's in the car).
> 
> 
> 
> So I wonder if I can use iBeacons? I confess to knowing very little about them and have a few questions I'd be really grateful if anyone can answer:
> 
> - Is there some good documentation about iBeacons (and in HA)? Also, how to use them with a Tesla.
> 
> - How does a Tesla use iBeacons? Does a Tesla continuously broadcast its presence using iBeacons and what information is in such broadcasts? If not continuously, what triggers its iBeacons? 
> 
> - I run HA on a RPi 4 and have a RaspBee II for ZigBee so I don't think using the Pi for BLE would be sensible, what would be the best alternative?
> 
> 
> 
> I know I'm asking for a repeat of some of the discussions above and in other threads but it's really hard to follow the approaches. It would be great to have a single "how to" article that could be updated as people do things differently. I'll certainly write up what I do and post it on my substack (free!) but at the moment I don't know where to start!

Please raise a new issue for this. I saw your request in the original thread and thought how it could be implemented. I'm not sure at the moment but it might be possible

**Comment by iainbullock** [comment]
> @iainbullock yes I found this special argument to be used on another forum, although I did not investigate what it does as "it works".
> 
> 
> 
> In your package shell simply add a special case for "wake" and change the simple $msg to "-domain vcsec $msg".
> 
> 
> 
> Indeed the --help gives lots of other possible commands. Basically LTE commands could really replace all the commands that fail through the Fleet API. Even the remote start, which Can be considered a security hazard in case of network/RPi hack.

I've added the suggested mod into the dev branch. I can't test it until I get back from holidays. Feel free to test it by using the dev tagged version in the Dockerhub repo

**Comment by m-chandler** [comment]
This seems to work for me on the dev branch.

1. Trying Honk when car is asleep (fails as expected)
```
tesla_ble_mqtt  | Received MQTT message: tesla_ble/command honk
tesla_ble_mqtt  | Command honk requested
tesla_ble_mqtt  | Honk Horn
tesla_ble_mqtt  | Sending command honk, attempt 1/5
tesla_ble_mqtt  | Error: context deadline exceeded
tesla_ble_mqtt  | Error calling tesla-control, exit code=1 - will retry in 5 seconds
tesla_ble_mqtt  | Sending command honk, attempt 2/5
tesla_ble_mqtt  | Error: ble: failed to enumerate device services: ATT request failed: input channel closed: io: read/write on closed pipe
tesla_ble_mqtt  | Error calling tesla-control, exit code=1 - will retry in 5 seconds
tesla_ble_mqtt  | Sending command honk, attempt 3/5
tesla_ble_mqtt  | Error: ble: failed to enumerate device services: ATT request failed: input channel closed: io: read/write on closed pipe
tesla_ble_mqtt  | Error calling tesla-control, exit code=1 - will retry in 5 seconds
tesla_ble_mqtt  | Sending command honk, attempt 4/5
tesla_ble_mqtt  | Error: context deadline exceeded
tesla_ble_mqtt  | Error calling tesla-control, exit code=1 - will retry in 5 seconds
tesla_ble_mqtt  | Sending command honk, attempt 5/5
tesla_ble_mqtt  | Error: context deadline exceeded
tesla_ble_mqtt  | Error calling tesla-control, exit code=1 - will retry in 5 seconds
```

2. Waking car (works as expected)
```
tesla_ble_mqtt  | Listening to MQTT
tesla_ble_mqtt  | Received MQTT message: tesla_ble/command wake
tesla_ble_mqtt  | Command wake requested
tesla_ble_mqtt  | Waking Car
tesla_ble_mqtt  | Sending command -domain vcsec wake, attempt 1/5
tesla_ble_mqtt  | Error: ble: failed to enumerate device services: ATT request failed: input channel closed: io: read/write on closed pipe
tesla_ble_mqtt  | Error calling tesla-control, exit code=1 - will retry in 5 seconds
tesla_ble_mqtt  | Sending command -domain vcsec wake, attempt 2/5
tesla_ble_mqtt  | Error: ble: failed to enumerate device services: ATT request failed: input channel closed: io: read/write on closed pipe
tesla_ble_mqtt  | Error calling tesla-control, exit code=1 - will retry in 5 seconds
tesla_ble_mqtt  | Sending command -domain vcsec wake, attempt 3/5
tesla_ble_mqtt  | Ok
```

3. Honk again now that car should be awake (works as expected)
```
tesla_ble_mqtt  | Listening to MQTT
tesla_ble_mqtt  | Received MQTT message: tesla_ble/command honk
tesla_ble_mqtt  | Command honk requested
tesla_ble_mqtt  | Honk Horn
tesla_ble_mqtt  | Sending command honk, attempt 1/5
tesla_ble_mqtt  | Ok
```

**Comment by raphmur** [comment]
Good, maybe close the issue? I think we are in line also on HA addon https://github.com/raphmur/tesla-local-control-addon/releases/tag/v0.0.8

**Comment by iainbullock** [comment]
Thanks everyone I will close this one
